### PR TITLE
Simplified API

### PR DESCRIPTION
**Pull request by kodebach**
DO NOT MERGE (won't compile anyway)

This PR is just for discussions and to have a nice diff.

This only includes changes that have been agreed upon. To propose new changes, create a comment in this PR.

### CODE REVIEW

File: src/include/kdb.h.in [hunk 1]
```diff
@@ -216,43 +188,31 @@ static inline Key *keyDup (const Key *source, elektraCopyFlags flags)
...
 
-ssize_t ksAppendKey (KeySet *ks, Key *toAppend);
+ElektraKeySet * elektraKeySetDup (const ElektraKeySet * source);
+int elektraKeySetCopy (ElektraKeySet * dest, const ElektraKeySet * source);
 
-ssize_t ksAppend (KeySet *ks, const KeySet *toAppend);
-KeySet *ksCut (KeySet *ks, const Key *cutpoint);
+uint16_t elektraKeySetIncRef (ElektraKeySet * ks);
+uint16_t elektraKeySetDecRef (ElektraKeySet * ks);
+uint16_t elektraKeySetGetRef (const ElektraKeySet * ks);
```
Comment (kodebach) [inline src/include/kdb.h.in]:
```suggestion
```

Should be moved to `libelektra-operations` together with `elektraKeyGetRef`. See also #4245.

File: src/include/kdb.h.in [hunk 2]
```diff
@@ -134,80 +125,61 @@ int kdbSet (KDB *handle, KeySet *returned,
...
-Key * keyCopy (Key *dest, const Key *source, elektraCopyFlags flags);
-
-int keyClear (Key *key);
-int keyDel (Key *key);
-
-uint16_t keyIncRef (Key *key);
-uint16_t keyDecRef (Key *key);
-uint16_t keyGetRef (const Key *key);
+ElektraKey * elektraKeyNew (const char * keyname, ...) ELEKTRA_SENTINEL;
+ElektraKey * elektraKeyVNew (const char * keyname, va_list ap);
```
Comment (kodebach) [inline src/include/kdb.h.in:129]:
```suggestion
ElektraKey * elektraKeyNew (elektraNamespace ns, const char * name, size_t nameSize);
```
see https://github.com/ElektraInitiative/libelektra/pull/4244

File: src/include/kdb.h.in [hunk 3]
```diff
@@ -216,43 +188,31 @@ static inline Key *keyDup (const Key *source, elektraCopyFlags flags)
...
-uint16_t ksIncRef (KeySet * ks);
-uint16_t ksDecRef (KeySet * ks);
-uint16_t ksGetRef (const KeySet * ks);
-
-int ksClear (KeySet *ks);
-int ksDel (KeySet *ks);
-
-ssize_t ksGetSize (const KeySet *ks);
+ElektraKeySet * elektraKeySetNew (size_t alloc, ...) ELEKTRA_SENTINEL;
+ElektraKeySet * elektraKeySetVNew (size_t alloc, va_list ap);
```
Comment (kodebach) [inline src/include/kdb.h.in]:
```suggestion
ElektraKeySet * elektraKeysetNew (ElektraKey ** keys, size_t keyCount);
```

see #4244

File: src/include/kdb.h.in [hunk 4]
```diff
@@ -216,43 +188,31 @@ static inline Key *keyDup (const Key *source, elektraCopyFlags flags)
...
 
-ssize_t ksAppendKey (KeySet *ks, Key *toAppend);
+ElektraKeySet * elektraKeySetDup (const ElektraKeySet * source);
+int elektraKeySetCopy (ElektraKeySet * dest, const ElektraKeySet * source);
```
Comment (kodebach) [inline src/include/kdb.h.in]:
```suggestion
```

Should be an optimization of `elektraKeySetAddAll` if `dest` is empty.

**Comment by kodebach** [comment]
Haven't had time yet to change this, but I just noticed the `elektraKeySet` prefix clashes with `elektraKeySet*` (i.e. a set operation on a key). We should use `elektraKeyset` instead.

File: src/include/kdb.h.in [hunk 5]
```diff
@@ -134,80 +125,61 @@ int kdbSet (KDB *handle, KeySet *returned,
...
 
-const char *keyBaseName (const Key *key);
-ssize_t keyGetBaseNameSize (const Key *key);
-ssize_t keyGetBaseName (const Key *key, char *returned, size_t maxSize);
+const char * elektraKeyEscapedName (const ElektraKey * key);
 
-ssize_t keySetBaseName (Key *key,const char *baseName);
-ssize_t keyAddBaseName (Key *key,const char *baseName);
+const char * elektraKeyBaseName (const ElektraKey * key);
+ssize_t elektraKeyBaseNameSize (const ElektraKey * key);
```
Comment (kodebach) [inline src/include/kdb.h.in:162]: Not really API, but the base name size should be stored in the struct as an optimization.

File: src/include/kdb.h.in [hunk 6]
```diff
@@ -216,43 +188,31 @@ static inline Key *keyDup (const Key *source, elektraCopyFlags flags)
...
-uint16_t ksIncRef (KeySet * ks);
-uint16_t ksDecRef (KeySet * ks);
-uint16_t ksGetRef (const KeySet * ks);
-
-int ksClear (KeySet *ks);
-int ksDel (KeySet *ks);
-
-ssize_t ksGetSize (const KeySet *ks);
+ElektraKeySet * elektraKeysetNew (size_t alloc, ...) ELEKTRA_SENTINEL;
+ElektraKeySet * elektraKeysetVNew (size_t alloc, va_list ap);
```
Comment (kodebach) [inline src/include/kdb.h.in:193]:
```suggestion
ElektraKeySet * elektraKeysetNew (ElektraKey ** keys, size_t keyCount);
```

see #4244

File: src/include/kdb.h.in [hunk 7]
```diff
@@ -216,43 +188,31 @@ static inline Key *keyDup (const Key *source, elektraCopyFlags flags)
...
 
-ssize_t ksAppendKey (KeySet *ks, Key *toAppend);
+ElektraKeySet * elektraKeysetDup (const ElektraKeySet * source);
+int elektraKeysetCopy (ElektraKeySet * dest, const ElektraKeySet * source);
```
Comment (kodebach) [inline src/include/kdb.h.in:196]:
```suggestion
```

This should be an optimization of `elektraKeysetAddAll` if `dest` is empty.

File: src/include/kdb.h.in [hunk 8]
```diff
@@ -216,43 +188,31 @@ static inline Key *keyDup (const Key *source, elektraCopyFlags flags)
...
 
-ssize_t ksAppend (KeySet *ks, const KeySet *toAppend);
-KeySet *ksCut (KeySet *ks, const Key *cutpoint);
+uint16_t elektraKeysetIncRef (ElektraKeySet * ks);
+uint16_t elektraKeysetDecRef (ElektraKeySet * ks);
+uint16_t elektraKeysetGetRef (const ElektraKeySet * ks);
```
Comment (kodebach) [inline src/include/kdb.h.in:200]:
```suggestion
```

should be moved to operations lib (see #4245)

File: src/include/kdb.h.in [hunk 9]
```diff
@@ -134,80 +125,61 @@ int kdbSet (KDB *handle, KeySet *returned,
...
+int elektraKeyClear (ElektraKey * key);
+int elektraKeyDel (ElektraKey * key);
 
-const Key *keyGetMeta (const Key *key, const char* metaName);
-ssize_t    keySetMeta (Key *key, const char* metaName,
-	const char *newMetaString);
-KeySet * keyMeta (Key * key);
+uint16_t elektraKeyIncRef (ElektraKey * key);
+uint16_t elektraKeyDecRef (ElektraKey * key);
+uint16_t elektraKeyGetRef (const ElektraKey * key);
```
Comment (kodebach) [inline src/include/kdb.h.in:138]:
```suggestion
```

should be moved to operations lib (see #4245)

File: src/include/kdb.h.in [hunk 10]
```diff
@@ -134,80 +125,61 @@ int kdbSet (KDB *handle, KeySet *returned,
...
 
-int keyIsBelow (const Key *key, const Key *check);
-int keyIsBelowOrSame (const Key *key, const Key *check);
-int keyIsDirectlyBelow (const Key *key, const Key *check);
-
-int keyIsBinary (const Key *key);
-int keyIsString (const Key *key);
+int elektraKeyIsBelow (const ElektraKey * key, const ElektraKey * check);
+int elektraKeyIsBelowOrSame (const ElektraKey * key, const ElektraKey * check);
+int elektraKeyIsDirectlyBelow (const ElektraKey * key, const ElektraKey * check);
```
Comment (kodebach) [inline src/include/kdb.h.in:149]:
```suggestion
```

should be moved to operations lib and merged (see #4245)

File: src/include/kdb.h.in [hunk 11]
```diff
@@ -216,43 +188,31 @@ static inline Key *keyDup (const Key *source, elektraCopyFlags flags)
...
 
-elektraCursor ksGetCursor (const KeySet *ks);
-int ksSetCursor (KeySet *ks, elektraCursor cursor);
-Key *ksAtCursor (KeySet *ks, elektraCursor cursor);
+ssize_t elektraKeysetAddAll (ElektraKeySet * ks, const ElektraKeySet * other);
 
-Key *ksLookup (KeySet *ks, Key *k, elektraLookupFlags options);
-Key *ksLookupByName (KeySet *ks, const char *name, elektraLookupFlags options);
+ElektraKey * elektraKeysetLookup (ElektraKeySet * ks, const ElektraKey * k);
+ElektraKey * elektraKeysetLookupByName (ElektraKeySet * ks, const char *name);
```
Comment (kodebach) [inline src/include/kdb.h.in:214]:
```suggestion
```

should be moved to operations lib (see #4245)

File: src/include/kdb.h.in [hunk 12]
```diff
@@ -134,80 +125,61 @@ int kdbSet (KDB *handle, KeySet *returned,
...
+ssize_t elektraKeyValueSize (const ElektraKey * key);
 
-ssize_t keyGetBinary (const Key *key, void *returnedBinary, size_t maxSize);
-ssize_t keySetBinary (Key *key, const void *newBinary, size_t dataSize);
+ssize_t elektraKeySetValue (ElektraKey * key, const void * value, size_t valueSize);
 
-int keyLock (Key * key, elektraLockFlags what);
-int keyIsLocked (const Key * key, elektraLockFlags what);
+/* Locking */
+int elektraKeyLock (ElektraKey * key, elektraLockFlags what);
```
Comment (kodebach) [inline src/include/kdb.h.in:177]: With the new automatic locking on keynames, it might make sense to have a separate function for that to highlight that locking of names works differently. But 3 separate functions also seems excessive.

File: src/include/kdb.h.in [hunk 13]
```diff
@@ -134,80 +125,61 @@ int kdbSet (KDB *handle, KeySet *returned,
...
+int elektraKeyIsDirectlyBelow (const ElektraKey * key, const ElektraKey * check);
 
 /* Name Manipulation Methods */
-const char *keyName (const Key *key);
-ssize_t keyGetNameSize (const Key *key);
-ssize_t keyGetName (const Key *key, char *returnedName, size_t maxSize);
-
-ssize_t keySetName (Key *key, const char *newname);
-ssize_t keyAddName (Key *key, const char *addName);
+const char * elektraKeyName (const ElektraKey * key);
```
Comment (kodebach) [inline src/include/kdb.h.in:152]: returns the unescaped name now

File: src/include/kdb.h.in [hunk 14]
```diff
@@ -134,80 +125,62 @@ int kdbSet (KDB *handle, KeySet *returned,
...
 
-/* Meta Info */
-int keyRewindMeta (Key *key);
-const Key *keyNextMeta (Key *key);
-const Key *keyCurrentMeta (const Key *key);
+ElektraKey * elektraKeyCopy (ElektraKey * dest, const ElektraKey * source, elektraCopyFlags flags);
 
-int keyCopyMeta (Key *dest, const Key *source, const char *metaName);
-int keyCopyAllMeta (Key *dest, const Key *source);
+int elektraKeyClear (ElektraKey * key);
```
Comment (kodebach) [inline src/include/kdb.h.in:133]:
```suggestion
```

should be moved to operations lib (see #4245)

**Note:** This is different from `elektraKeyCopy (key, NULL, ELEKTRA_KEY_CP_ALL)`. The intention is that this snippet

```c
elektraKeyClear (k);
```

is the same as this 

```c
if (elektraKeyDel (k) == 0) {
    k = elektraKeyNew ("/", ELEKTRA_KEY_END);
}
```

but without the new allocation. This could be useful in (some) loops to avoid repeated re-allocation. `elektraKeyCopy` would only clear name, value and metadata and also respects locks.

File: src/include/kdb.h.in (hunk 5 above)
Comment (markus2330) [reply src/include/kdb.h.in:162]: No, without extensive benchmarks that this is really something that pays off (which I doubt) this is a bad idea (classical premature optimization).

Comment (kodebach) [reply src/include/kdb.h.in:162]: I assumed that we would remove fields from `struct _Key`. If the space gained there is used for the base name size, we don't loose anything in terms of memory. In terms of performance, we may not gain much, but storing the size is definitely not slower.

> extensive benchmarks [...] something that pays off (which I doubt)

It is very hard to say, how often the base name size is needed by real world use cases. This makes constructing valid benchmarks very hard.

`elektraKeyBaseNameSize` is probably almost never called directly. However, if the size were know, it could be used in `elektraKeyBaseName` and `elektraKeySetBaseName` (instead of iterating backwards). Those functions are definitely used more (how much is uncertain).

One thing however is certain: There is a (albeit small) performance gain when switching from:

```c
const char * baseName = key->ukey + key->keyUSize - 2;
while (*baseName != '\0')
{
	--baseName;
}
return baseName + 1;
```
to just
```c
return key->ukey + key->keyUSize - key->baseNameSize;
```

The remaining question is: Is this snippet (or the equivalent in `elektraKeySetBaseName`) called often enough that this difference adds up to something significant?

However, once again, this question is only really relevant, if there is a possible downside to adding `key->baseNameSize`.

File: src/include/kdb.h.in [hunk 15]
```diff
@@ -46,58 +46,51 @@ extern "C" {
...
-	KEY_FLAGS=3,
-	KEY_BINARY=1<<4,
-	KEY_SIZE=1<<11,
-	KEY_FUNC=1<<12,
-	KEY_META=1<<15,
-	KEY_NULL=1<<16,
-	KEY_LOCK_NAME=1<<17,
-	KEY_LOCK_VALUE=1<<18,
-	KEY_LOCK_META=1<<19,
+	ELEKTRA_KEY_KEY_VALUE=1<<1,
```
[the remark this answers is not in the document]
Comment (kodebach) [reply src/include/kdb.h.in:49]: Not sure what happend there... It was supposed to be `ELEKTRA_KEY_VALUE`, etc.

(Assuming we keep those constants at all)

File: src/include/kdb.h.in [hunk 16]
```diff
@@ -107,9 +100,9 @@ namespace ckdb {
 extern "C" {
 #endif
 
-typedef struct _KDB	KDB;
-typedef struct _Key	Key;
-typedef struct _KeySet	KeySet;
+typedef struct _KDB	ElektraKDB;
```
[the remark this answers is not in the document]
Comment (kodebach) [reply src/include/kdb.h.in:103]: 1. How would you redefine this, without rewriting the entire header file? As soon as you include the header there will be a conflict, if there is already a `KDB`. This will cause a compile error.
2. I think it is easier for us and for users, if we just consistently prefix _everything_ with some form of `Elektra`. Then you know this belongs to Elektra when reading code, we avoid all sorts of conflicts and we don't need to think whether or not to use the prefix.

File: src/include/kdb.h.in [hunk 17]
```diff
@@ -134,80 +125,62 @@ int kdbSet (KDB *handle, KeySet *returned,
...
-KeySet * keyMeta (Key * key);
+uint16_t elektraKeyIncRef (ElektraKey * key);
+uint16_t elektraKeyDecRef (ElektraKey * key);
+uint16_t elektraKeyGetRef (const ElektraKey * key);
 
-/* Methods for Making Tests */
-int keyCmp (const Key *k1, const Key *k2);
+/* Metadata */
+ElektraKeySet * elektraKeyMeta (ElektraKey * key);
+int elektraKeySetMeta (ElektraKey * key, ElektraKeySet * meta);
```
Comment (lawli3t) [inline src/include/kdb.h.in:142]: One thing I noticed while reading this, is that it might get misconstrued as KeySet Meta

Comment (kodebach) [reply src/include/kdb.h.in:142]: Yes, I noticed that too. That's why I called all the keyset functions `elektraKeyset*`. Probably, the struct should also be called `ElektraKeyset` to avoid confusion.

Comment (lawli3t) [reply src/include/kdb.h.in:142]: Yes, I just noticed as I was updating the Keyset functions. At least alleviates the issues, renaming the struct is a good idea as well. I tried to think of better names, but this would just make everything inconsistent again I guess.

Comment (markus2330) [reply src/include/kdb.h.in:142]: Yes, if we prefix with elektra we also need to rename Keyset. Alternative would be KS, resembling KDB.

File: src/include/kdb.h.in (hunk 9 above)
Comment (markus2330) [reply src/include/kdb.h.in:138]: To be discussed in #4245 where it belongs best. Maybe we need some diagnostic lib.

File: src/include/kdb.h.in (hunk 10 above)
Comment (markus2330) [reply src/include/kdb.h.in:149]: I agree about the merging, not sure if it can be moved to operations (cyclic links to ksCut)

File: src/include/kdb.h.in (hunk 5 above)
Comment (markus2330) [reply src/include/kdb.h.in:162]: The problem of invalid benchmarks is much smaller than of invalid assumptions in the guesswork of what might be faster.

I don't see a need to have the size stored and without conclusive data that shows the need we shouldn't add it as we already know we are too memory-intensive. So actually we also shouldn't provide API that returns it.

**Review by markus2330 — CHANGES_REQUESTED** [review]
Great job! I close this now as the header file is part of #4341

File: src/include/kdb.h.in [hunk 18]
```diff
@@ -118,14 +111,12 @@ typedef struct _KeySet	KeySet;
  *
  **************************************/
 
-KDB * kdbOpen (const KeySet * contract, Key *parentKey);
+ElektraKDB * elektraKdbOpen (const ElektraKeySet * contract, ElektraKey * parentKey);
 
-int kdbClose (KDB *handle, Key *errorKey);
+int elektraKdbClose (ElektraKDB * handle, ElektraKey * errorKey);
```
[the remark this answers is not in the document]
Comment (kodebach) [reply src/include/kdb.h.in:116]: Actually, now that I look at it, it should probably be

```suggestion
int elektraKdbClose (ElektraKdb * handle, ElektraKey * errorKey);
```

I prefer the consistent application of casing rules. With the `camelCase` rules, you write the first letter of every word/term in uppercase and everything in lowercase. Using the rules consistently also allows a direct translation to other casing rules when required (e.g. when naming a constant in `SCREAMING_SNAKE_CASE`).

The direct translation of `ElektraKDB` from `PascalCase` to `SCREAMING_SNAKE_CASE` would be `ELEKTRA_K_D_B`.

File: src/include/kdb.h.in (hunk 10 above)
Comment (kodebach) [reply src/include/kdb.h.in:149]: IMO this is more a question of "Is `ksCut` minimal enough for the core?"

File: src/include/kdb.h.in (hunk 18 above)
[the remark this answers is not in the document]
Comment (markus2330) [reply src/include/kdb.h.in:116]: I also prefer to consistent use of casing. Acronyms, however, have different rules and are usually written in all-capital letters in the English language. So we need a rule for how to deal with acronyms anyway.

> ```suggestion
> int elektraKdbClose (ElektraKdb * handle, ElektraKey * errorKey);
> ```

Looks actually also nice.

> `ELEKTRA_K_D_B`

Imho, not so pretty.

Can you please write a decision and/or update DESIGN.md for the casing and acronyms?

Do we use any other acronym next to Kdb?

File: src/include/kdb.h.in (hunk 10 above)
Comment (markus2330) [reply src/include/kdb.h.in:149]: If ksCut is the only thing that depends then case is quite clear: then we move elektraKeyIsBelow and elektraKsCut (or elektraKeysetCut?) to the operations lib.

File: src/include/kdb.h.in (hunk 5 above)
Comment (markus2330) [reply src/include/kdb.h.in:162]: > IMO making reasonable assumptions is much better than making confident mistakes based on artificial benchmarks. If your benchmarks don't reflect the real world, basing decisions off of those benchmarks might make things a lot worse.

Making things more objective (or said differently: working more scientific) per se cannot "make things a lot worse". If we get too confident about data that actually does not provide this kind of confidence (i.e. always), we have a similar but not identical problem as when we simply confidently claim something without any backup (i.e. pure opinion).

> Of course the best middle ground would be to make assumptions about the real world and build benchmarks based on that. However, benchmarks for the this example are actually really hard to design. You need to know how often the basename is read/updated to know how much speedup/slowdown would be noticeable. But you also need to know how big basenames are, since the current implementation has runtime O(size of basename).

Kurt Micheli made some very good work in the direction of generating KeySets with different basename lengths. If your benchmark can show improvements across many different basename sizes/KeySet shapes, the "realism" of how the KeySet looks like gets less important. See benchmarks/opmphm.c

Comment (kodebach) [reply src/include/kdb.h.in:162]: I don't think this discussion will lead anywhere. It is clear to me that you are absolutely insistent on having benchmarks before changing anything, even if the change can be clearly doesn't impact memory use (same `struct` size) and is from a `O(size of basename)` to a `O(1)` runtime.

I assume this is because, there is the option to reduce the memory footprint (by removing the second size field from the struct) and keeping the current performance. But that means we need a benchmark must show whether the increased performance of the `O(1)` implementation is worth the increased memory footprint.

Such a trade-off can, however, only be evaluated with real-world data. Without the real-world data you can only try all possible combinations of basename size, and number of calls to the various `elektraKey*BaseName*` functions. This will generate huge amounts of data and then what? It's still a trade-off between memory and runtime you can't decided that without real world data.

> See benchmarks/opmphm.c

That is a totally different problem. That benchmark compares the runtime of two implementations. We could do this here too, but I don't see the point since one implementation is `O(1)` (actually constant in all cases and not just mostly constant like a hashmap) and the other is not.

---

In summary, I think we can leave it for now and not store the basename size. In future we can revisit this and maybe then we have some real world data.
